### PR TITLE
Added extract archive button to toolbar

### LINKED ISSUE #7639: Show an Extract all button when selecting a .zip file
### What's the Problem?

Just like in the Windows Explorer application, make a button appear with the text Extract all when clicking once on a .zip file

### Solution/Idea

Add a button to extract all contents from a zip file 

### Alternatives

Maybe A shortcut like CTRL + E to extract

### Priorities

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to click A button in the icon bar that extracts all contents from a .zip file | Must |
| This proposal will allow developers to click a shortcut (maybe CTRL + E) to extract all contents from a zip file | Could |

### Files Version

_No response_

### Windows Version

_No response_

### Comments

_No response_

### PR DESCRIPTION
**Pull request by puppetsw**
When selecting a zip archive, the extract button will show on the toolbar. Shortcut key of CTRL+E has also been added.

**Resolved / Related Issues**
- Closes #7639

**Details of Changes**
- Added extract button in toolbar when clicking an archive file.
- Added shortcut key of CTRL+E (enabled when button shows)

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [x] Tested the changes for accessibility

**Screenshots (optional)**
Example12

**Thoughts**
- Should the button text be shown? It may look more consistent without the text.

### CODE REVIEW

**Comment by yaichenbaum** [comment]
@puppetsw you can resolve the merge conflicts?

File: src/Files/UserControls/InnerNavigationToolbar.xaml [hunk 1]
```diff
@@ -190,6 +190,26 @@
                         <FontIcon FontFamily="{StaticResource RecycleBinIcons}" Glyph="&#xEF88;" />
                     </AppBarButton.Icon>
                 </AppBarButton>
+                <AppBarButton
```
Comment (yaichenbaum) [inline src/Files/UserControls/InnerNavigationToolbar.xaml:193]: Can you add a separator that's shows when the extract button is visible?

Comment (puppetsw) [reply src/Files/UserControls/InnerNavigationToolbar.xaml:193]: Yep, I'll do that and fix the merge conflicts.

File: src/Files/UserControls/InnerNavigationToolbar.xaml [hunk 2]
```diff
@@ -205,6 +205,27 @@
...
+                    x:Name="ExtractButton"
+                    Width="Auto"
+                    MinWidth="40"
+                    Command="{x:Bind ViewModel.ExtractCommand, Mode=OneWay}"
+                    IsEnabled="{x:Bind ViewModel.CanExtract, Mode=OneWay, FallbackValue=False}"
+                    Label="{x:Bind GetLocalizedString('BaseLayoutItemContextFlyoutExtract/Text')}"
+                    LabelPosition="Default"
+                    Visibility="{x:Bind ViewModel.CanExtract, Mode=OneWay, FallbackValue=Collapsed}"
+                    KeyboardAcceleratorTextOverride="Ctrl+E">
```
Comment (yaichenbaum) [inline src/Files/UserControls/InnerNavigationToolbar.xaml:218]: Do you want to add a resource string for this?

Comment (puppetsw) [reply src/Files/UserControls/InnerNavigationToolbar.xaml:218]: Sure I can do that. Just want make sure I'm doing it correct though. I'll go through each resource file and add a "Extract" string to each and change the xaml to be 
```xml 
Label="{helpers:ResourceString Name=Extract}" 
```

Comment (yaichenbaum) [reply src/Files/UserControls/InnerNavigationToolbar.xaml:218]: That's correct, if you're up to it, you can remove duplicate strings for "Extract" and use the new one instead.

**Comment by yaichenbaum** [comment]
@puppetsw do you think it would be useful to display a flyout menu to provide additional options?
image

File: src/Files/UserControls/InnerNavigationToolbar.xaml (hunk 2 above)
Comment (puppetsw) [reply src/Files/UserControls/InnerNavigationToolbar.xaml:218]: Excellent, that sounds like something I can do. I’ll get on it.

**Comment by puppetsw** [comment]
@yair100 the flyout could work. I’ll experiment with it. The original intention was to provide a quick way to extract the file. I just tried to emulate what was suggested in the issue. The Win11 file explorer works similarly with the single button.

File: src/Files/UserControls/InnerNavigationToolbar.xaml (hunk 2 above)
Comment (yaichenbaum) [reply src/Files/UserControls/InnerNavigationToolbar.xaml:218]: You can also add a resource string called `ExtractKeyboardAcceleratorTextOverride`

**Comment by yair100** [comment]
> @yair100 the flyout could work. I’ll experiment with it. The original intention was to provide a quick way to extract the file. I just tried to emulate what was suggested in the issue. The Win11 file explorer works similarly with the single button.

We can merge the pull request without the flyout but I think it will be a nice improvement over the implementation in File Explorer.

**Comment by puppetsw** [comment]
> > @yair100 the flyout could work. I’ll experiment with it. The original intention was to provide a quick way to extract the file. I just tried to emulate what was suggested in the issue. The Win11 file explorer works similarly with the single button.
> 
> We can merge the pull request without the flyout but I think it will be a nice improvement over the implementation in File Explorer.

I've added the resource strings. I'll work on the flyout tomorrow.

File: src/Files/Strings/da/Resources.resw [hunk 3]
```diff
@@ -1686,4 +1791,10 @@
   <data name="InsertDiscDialog.Title" xml:space="preserve">
     <value>Indsæt en disk</value>
   </data>
+  <data name="Extract" xml:space="preserve">
+    <value>Udpak</value>
+  </data>
+  <data name="ExtractKeyboardAcceleratorTextOverride" xml:space="preserve">
+    <value>Ctrl+E</value>
+  </data>
```
Comment (yaichenbaum) [inline src/Files/Strings/da/Resources.resw:1799]: Can you revert the changes to the resw files? We usally generate them separately to make the PR review easier.

**Comment by puppetsw** [comment]
I've changed it to a flyout, but I wasn't sure the best way of handling the path for the "Extract to" button. I've got it bound to a property in the ViewModel at the moment.

Example12

**Comment by yaichenbaum** [comment]
What do you think about adding back the text to extract button?

File: src/Files/Strings/af/Resources.resw [hunk 4]
```diff
@@ -2538,4 +2643,10 @@ Ons gebruik App Center om tred te hou met programgebruik, foute te vind en ongel
   <data name="SettingsSetAsDefaultFileManagerDescription" xml:space="preserve">
     <value>Hierdie instelling verander stelsellêers en kan onverwagte newe-effekte op jou toestel hê. Die ontwikkelaars neem geen verantwoordelikheid in die geval dat 'n probleem as gevolg daarvan plaasvind nie. Om voort te gaan met hierdie opsie is 'n erkenning van die risiko's wat by hierdie aksie betrokke is. Let asseblief daarop dat die deïnstalleer van Files nie hierdie veranderinge sal ongedaan maak nie en kan verhoed dat jy File Explorer oopmaak, tensy jy hierdie instelling afskakel voordat jy Files van jou toestel verwyder.</value>
   </data>
+  <data name="Extract" xml:space="preserve">
+    <value>Uitreksel</value>
+  </data>
+  <data name="ExtractKeyboardAcceleratorTextOverride" xml:space="preserve">
+    <value>Ctrl+E</value>
+  </data>
```
Comment (yaichenbaum) [inline src/Files/Strings/af/Resources.resw:2651]: Can you revert the changes to the resw files?

Comment (puppetsw) [reply src/Files/Strings/af/Resources.resw:2651]: I believe the resw files should be reverted now.

**Comment by puppetsw** [comment]
> What do you think about adding back the text to extract button?

Yep, I can add that back in.

**Comment by yaichenbaum** [comment]
@puppetsw it's working well! One little nitpick, if the menu is open the user clicks away from the selected item, the extract button becomes hidden but the flyout stays open. Can you see if it's possible to improve this behavior?

**Comment by puppetsw** [comment]
> @puppetsw it's working well! One little nitpick, if the menu is open the user clicks away from the selected item, the extract button becomes hidden but the flyout stays open. Can you see if it's possible to improve this behavior?

I’ll investigate and see if it can be fixed.

**Comment by puppetsw** [comment]
> @puppetsw it's working well! One little nitpick, if the menu is open the user clicks away from the selected item, the extract button becomes hidden but the flyout stays open. Can you see if it's possible to improve this behavior?

I believe this issue is now fixed. 

Just one thing I noticed was that the "Extract to ..." option is missing the ending \ that is shown on the other extract context menu. Just think it should be added, or removed from the other for consistency.

**Comment by yaichenbaum** [comment]
@puppetsw can you add the slash for consistency?

**Comment by cinqmilleans** [comment]
@puppetsw You shouldn't add ExtractButtonSeparator but use the generic AdditionalActionSeparator.
You just have to modify in the viewmodel HasAdditionnalAction and UpdateAdditionnalActions.

**Comment by puppetsw** [comment]
> @puppetsw You shouldn't add ExtractButtonSeparator but use the generic AdditionalActionSeparator. You just have to modify in the viewmodel HasAdditionnalAction and UpdateAdditionnalActions.

The UpdateAdditionalActions method doesn't seem to be called when the selection changes, but only when the page changes. I could add it to the SelectedItems property. What are your thoughts?

**Comment by cinqmilleans** [comment]
@puppetsw Ok, if you need it you have to do it.

**Review by yair100 — DISMISSED** [review]

**Comment by puppetsw** [comment]
@yaichenbaum Sorry, didn't realise you were working on the conflicts

File: src/Files/ViewModels/NavToolbarViewModel.cs [hunk 5]
```diff
@@ -1080,17 +1086,38 @@ public List<ListedItem> SelectedItems
...
                 }
             }
         }
 
-        public bool HasAdditionnalAction => InstanceViewModel.IsPageTypeRecycleBin;
+        public bool HasAdditionnalAction
+        {
+            get
+            {
+                if (InstanceViewModel.IsPageTypeRecycleBin)
```
[the remark this answers is not in the document]
Comment (puppetsw) [reply src/Files/ViewModels/NavToolbarViewModel.cs]: Updated

File: src/Files/Helpers/MenuFlyoutHelper.cs [hunk 6]
```diff
@@ -59,6 +59,30 @@ internal MenuFlyoutSubItemViewModel(string text) : base(text)
...
+        
+        public static bool GetIsVisible(DependencyObject d) => (bool)d.GetValue(IsVisibleProperty);
+
+        public static void SetIsVisible(DependencyObject d, bool value) => d.SetValue(IsVisibleProperty, value);
+
+        public static readonly DependencyProperty IsVisibleProperty = DependencyProperty.RegisterAttached("IsVisible", typeof(bool), typeof(MenuFlyoutHelper), new PropertyMetadata(false, OnIsVisiblePropertyChanged));
+
+        private static void OnIsVisiblePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            var flyout = d as MenuFlyout;
```
[the remark this answers is not in the document]
Comment (puppetsw) [reply src/Files/Helpers/MenuFlyoutHelper.cs]: Updated

**Review by yair100 — APPROVED** [review]
